### PR TITLE
Prevent ID collisions in posts queries that span multiple sites

### PR DESCRIPTION
**Pull request by nylen**
I found this issue while investigating #17549 and found it necessary to fix it to test that change sufficiently.

I created a pair of test sites `test1user2sites` and `test1user2sites2`, both owned by the user `test1user2sites`.  Because I created each one using the same steps, they have the same post IDs assigned to their first two posts:  4 and 16.

When rendered into a `PostTypeList`, this is an ID collision, and entries with the same ID are only pulled from 1 site:

We can fix this by using `global_ID` as the key for `PostQueryManager`.  Due to uncertainty about the broader effects of making this change for all posts queries, I have only changed the key for queries that involve multiple sites.

### To test

Log in as user `test1user2sites`.

Ensure that the changes from #17549 are applied, whether after its merge or through other means.

Start Calypso using `ENABLE_FEATURES=posts/post-type-list npm start`.

Visit `/posts` and verify that before this PR, the posts list displays broken, duplicated posts as in the first screenshot, and after, the posts from both sites are displayed correctly as in the second screenshot.

### CODE REVIEW

**Comment by nylen** [comment]
@aduth I think you may be interested in this one...

**Comment by nylen** [comment]
Thanks.  Setting back to "In Progress" as this needs further exploration of the points above.

**Comment by nylen** [comment]
The last two commits change the query key for all `PostQueryManager` instances to `global_ID`.

If this path is followed, it will end up being an extensive refactor touching many other actions and parts of the state tree.  Here's one example of why:  the [`POST_RESTORE` action](https://github.com/Automattic/wp-calypso/blob/988fcc8/client/state/posts/reducer.js#L162-L167) accepts a site and post ID, and `QueryManager` [expects to use this post ID as its item key](https://github.com/Automattic/wp-calypso/blob/988fcc8/client/lib/query-manager/index.js#L234) when processing this action via the `receive` method.

So, basically all actions that take a site and post ID would need to be refactored to use a global post ID instead.

I don't think this is a great path forward:  many of these actions really do apply to a specific, single site and a specific, single post ID.

Instead, I'm going to spend a bit more time today looking into moving multi-site queries out to a separate part of the state tree.  I'm not very concerned about the branching logic here, because it's fine for single-site posts queries to be keyed by numeric post ID, and multi-site posts queries are inherently very different, so can be handled differently.

**Comment by nylen** [comment]
> I'm going to spend a bit more time today looking into moving multi-site queries out to a separate part of the state tree

This isn't going to be so easy either:  we'll need to copy all the actions that we expect to take place for all-sites queries out to a separate reducer.  This could still be a cleaner path forward, as the current code doesn't make an explicit distinction between queries for a single site vs. queries for all sites.

I also tried moving multi-site queries to "site ID" `0`.  There are a lot of places in the code that pass `null` as a site ID, and it looks like again this is a mix between intentionally using `null` to mean all sites, and passing `null` because a site ID value is implicitly assumed to be present but not available.

For now, I've left this PR with a more explicit version of the initial code, with a warning if an attempt is made to fetch a post with siteId `null` and a numeric post ID.

I'm able to trigger this warning using the following steps:

- Visit `/posts`
- Click a post title to edit that post
- Wait for the post to load, then click the Back icon at the top left of the page (or the browser Back button, I see the same result either way).

2017-09-29t18 52 42-0500

> Schema seems fine to handle (specifies `array` type for item keys), but I noticed it keys queries by numeric site ID. Is this working for the null site?

This was indeed an issue.  With the current state structure it's fixed in 657e2c3.

**Comment by nylen** [comment]
We've decided on an approach here:  moving multi-site queries out to a separate part of the state tree.  I'll close this PR and begin this work in a new PR.
